pkg/agent/controller/networkpolicy/reject.go

```diff
@@ -153,13 +159,34 @@ func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
 		gwIfaces := c.ifaceStore.GetInterfacesByType(interfacestore.GatewayInterface)
 		return dstIsLocal && dstMAC == gwIfaces[0].MAC.String()
 	}
-	packetOutType := getRejectType(isServiceTraffic(), c.antreaProxyEnabled, srcIsLocal, dstIsDirect)
+
+	// When rejecting external client access to a Service with a remote Endpoint, there are two scenarios:
+	// 1. If the remote Endpoint is reachable via the OVS tunnel interface (i.e., traffic mode is "encap" or "hybrid"
+	//    with the remote Node on a different subnet), the packet-out should simulate being sent from the tunnel.
+	// 2. If the remote Endpoint is reachable via the OVS gateway interface and Node route (i.e., traffic mode is
+	//    "noEncap" or "hybrid" with the remote Node on the same subnet), the packet-out should simulate being sent
+	//    from the gateway.
+	// The PktDestinationField set in L3Forwarding table indicates the intended outPort of the packet-in packet, which
+	// can help determine the appropriate inPort for the packet-out packet.
+	var srcFromTun bool
+	if match := getMatchRegField(matches, openflow.PktDestinationField); match != nil {
+		pktDestinationVal, err := getInfoInReg(match, openflow.PktDestinationField.GetRange().ToNXRange())
+		if err != nil {
+			return err
+		}
+		srcFromTun = pktDestinationVal == openflow.ToTunnelRegMark.GetValue()
+	}
+
+	packetOutType := getRejectType(isServiceTraffic(), c.antreaProxyEnabled, srcIsLocal, dstIsDirect, srcFromTun)
 	if packetOutType == unsupported {
 		return fmt.Errorf("error when generating reject response for the packet from: %s to %s: neither source nor destination are on this Node", dstIP, srcIP)
 	}
-	if packetOutType == rejectServiceRemoteToExternal {
+	if packetOutType == rejectServiceRemoteFromTunToExternal {
```

Comment:
@hongliangl I might misunderstand something, could you help me understand why we are setting the response's dst MAC to the virtual MAC? I thought the packet should be sent to external client via antrea-gw0, no?

Reply:
>  I thought the packet should be sent to external client via antrea-gw0

You are right. The packet should be sent to external client via antrea-gw0 after they are processed by OVS pipeline. 

The packet-out is to stimulate a packet before it is processed by OVS pipeline. To stimulate a reject packet from remote Node via tunnel, the dst MAC should be the virtual MAC since it is received on the tunnel.